### PR TITLE
Enumerables

### PR DESCRIPTION
**Pull request by MarkoNS1990**
🟢  In this project we were making a clone of enumerable methods such as #each, #each_with_index, #select, #all?, #none?, #any?, #count, #inject, #map 
We wrote map by accepting both proc and block as an arguments. 🟢

### CODE REVIEW

**Comment by mimipeshy** [comment]
### STATUS: CHANGES REQUIRED :writing_hand:
Hello team, a great effort so far. 👏🏾 

#### To Highlight

- [ ] Good job on following the correct GitHub flow 👌🏾 

#### To Improve

- You have done a nice job implementing the enumerables but there are a few fixes you have to implement in order for this project to get approved.

- [x] Your README file has a lot of errors. Copy the contents and paste them into Grammarly to help you with the errors
- [ ] Commit messages should always be written in the present tense and should start with a capital letter eg `Add Proc Method`. You can change the ones written in the past tense using `git rebase`
- [x] enumerables #my_each returns an Enumerator if no block is given
- [x] enumerables #my_each when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- [x] enumerables #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [x] enumerables #my_each when a block is given when self is a range behaves like #each calls the given block once for each element in self
- [x]  enumerables #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self
- [x]  enumerables #my_each_with_index returns an enumerator if no block is given
- [x]  enumerables #my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- [x] enumerables #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [x] enumerables #my_each_with_index when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
- [x] enumerables #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self
- [x]  enumerables #my_select returns an array containing all elements of enum for which the given block returns a true value::range
- [x]  enumerables #my_select returns an enumerator if no block is given
- [x]  enumerables #my_all? returns false if the block returns false or nil::range
- [x]  enumerables #my_all? when no block or argument is given returns true when none of the collection members are false or nil
- [x]  enumerables #my_all? when no block or argument is given returns false when one of the collection members are false or nil
- [x] enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
- [x] enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- [x] enumerables #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
- [x] enumerables #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] enumerables #my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex
- [x]  enumerables #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [x] enumerables #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
- [x] enumerables #my_any? returns true if the block ever returns a value other than false or nil
- [x] enumerables #my_any? returns false if the block ever returns a value false or nil
- [x] enumerables #my_any? returns false if the block ever returns a value false or nil::range
- [x]  enumerables #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
- [x]  enumerables #my_any? when no block or argument is given returns false if at least one of the collection is not true
- [x] enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
- [x] enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
- [x] enumerables #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
- [x]  enumerables #my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x]  enumerables #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- [x] enumerables #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
- [x] enumerables #my_none? returns false if the block ever returns true for all elements::range
- [x]  enumerables #my_none? when no block or argument is given returns true only if none of the collection members is true
- [x] enumerables #my_none? when no block or argument is given returns false only if one of the collection members is true
- [x] enumerables #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
- [x] enumerables #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
- [x]  enumerables #my_none? when a class is passed as an argument returns false if any of the collection is a member of such class
- [x] enumerables #my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
- [x] enumerables #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
- [x]  enumerables #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- [x] enumerables #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
- [x] enumerables #my_count returns the number of items in enum through enumeration
- [x] enumerables #my_count returns the number of items in enum through enumeration::range
- [x]  enumerables #my_count counts the number of items in enum that are equal to item if an argument is given
- [x] enumerables #my_map returns a new array with the results of running a given block once for every element in en
- [x] enumerables #my_map returns a new array with the results of running a given block once for every element in enum
- [x]  enumerables #my_map returns a new array with the results of running a given block once for every element in enum::range
- [x]  enumerables #my_map returns an Enumerator if no block is given
- [x] enumerables #my_map accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.
- [x] enumerables #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
- [x] enumerables #my_map does not mutate the original array
- [x] enumerables #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
- [x]  enumerables #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
- [x] enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
- [x] enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
- [x] enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
- [x] enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range

That's all :smile: kindly submit another review when done.

_Feel free to reach out to me on slack if you have any questions or something is unclear, or just comment on this PR and tag my name  (@mimipeshy) and I will be glad to help_
Regards
Peris Ndanu

**Review by GzimAsani — CHANGES_REQUESTED** [review]
# INVALID PULL REQUEST :red_circle: 

### Hello team, I will invalidate this pull request because your linters are failing. So please fix them and ask for another review

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,166 @@
+# rubocop: disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity,Metrics/ModuleLength
+module Enumerable
+  def my_each
```
Comment (GzimAsani) [inline enumerables.rb:3]: 1. When a block is given when self is a range returns the range itself after calling the given block once for each element in self
_NameError:
       undefined local variable or method `length' for 5..50:Range_

2. When a block is given when self is a range behaves like #each calls the given block once for each element in self
_NameError:
       undefined local variable or method `length' for 5..50:Range_
3. When a block is given when self is a hash behaves like #each calls the given block once for each element in self
_expected: "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"
            got: ""_

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    i = 0 # iterator
+    while i < length
+      yield self[i]
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
```
Comment (GzimAsani) [inline enumerables.rb:14]: 1. when a block is given when self is a range returns the range itself after calling the given block once for each element in self
_NameError:
       undefined local variable or method `length' for 5..50:Range_
2. when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
_NameError:
       undefined local variable or method `length' for 5..50:Range_
3. when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+    i = 0
+    while i < length
+      yield self[i], i
+      i += 1
+    end
+    self
+  end
+
+  def my_select
```
Comment (GzimAsani) [inline enumerables.rb:25]: 1. Returns an array containing all elements of enum for which the given block returns a true value::range
_NameError:
       undefined local variable or method `length' for 5..50:Range_

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    new_arr = []
+    i = 0
+    while i < length
+      new_arr << self[i] if yield self[i]
+      i += 1
+    end
+    new_arr
+  end
+
+  def my_all?(item1 = nil)
```
Comment (GzimAsani) [inline enumerables.rb:37]: 1. Returns false if the block returns false or nil::range
_NameError:
       undefined local variable or method `length' for 5..50:Range_

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      return true
+    elsif !item1.nil? && item1.instance_of?(Regexp)
+      my_each { |i| return false unless item1.match(i) }
+    else
+      my_each { |i| return false if i != item1 }
+    end
+    true
+  end
+
+  def my_any?(item1 = nil)
```
Comment (GzimAsani) [inline enumerables.rb:54]: 1. returns false if the block ever returns a value false or nil::range
_NameError:
       undefined local variable or method `length' for 5..50:Range_

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+      my_each { |i| return true if i.is_a?(item1) }
+    elsif !item1.nil? && item1.instance_of?(Regexp)
+      my_each { |i| return true if item1.match(i) }
+    else
+      my_each { |i| return true if i == item1 }
+    end
+    false
+  end
+
+  def my_none?(item1 = nil)
```
Comment (GzimAsani) [inline enumerables.rb:70]: 1. returns false if the block ever returns true for all elements::range
_NameError:
        undefined local variable or method `length' for 5..50:Range_

2. when no block or argument is given returns false only if one of the collection members is true
_expected false
             got true_

3. when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
_expected false
             got true_

**Review by GzimAsani — CHANGES_REQUESTED** [review]
# STATUS: Changes Required :red_circle: 

### To highlight:
- Great job, implementing most of the requirements asked from the previous reviewers

### To improve:
- [ ] Your methods are showing 12 errors which are coming from `#my_each?`, `#my_each_with_index?`, `my_select `, `#my_all?`, `#my_any?`, `#my_none?`, check the comments below for more information:

Please fix the required changes and ask for another code review whenever you're ready

_**If you have any question related to the required changes feel free comment here, or ping me on slack @GzimAsani**_

HAPPY CODING :computer: :desktop_computer:

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+# # my_count
+# count = [12, 16, 17, 18, 20, 22, 23].my_count { |num| num <17 } # 4
+# puts count
+# # my_map
+# new_arr2 = [13, 15, 18, 19].my_map { |num| num + 1 } # [14,16,19,20]
+# p new_arr2
+# # my_inject
+# [1, 2, 3].my_inject { |sum, num| sum + num } # result = 6
+# [1,2,3].my_inject # error - no block given
+# puts multiply_els([2, 4, 5]) # 40
```
Comment (enaburekhan) [inline enumerables.rb]: It’s better to delete commented pieces of codes. If you need to review a previous version of your code, use git history.

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:full_moon:
**Hi team!
Great job so far. Thank you for implementing most of the changes requested by the last reviewer:clap::clap:. We are almost there. 
We just have two tests in  #my_none that are currently failing. Let's quickly fix it and get this project approved:muscle:.**

**To improve**

- [ ] #my_none? when no block or argument is given returns false only if one of the collection members is true. Expected false, instead, it's returning true. Currently, the method does not cover this test. Kindly add it.

- [ ] #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collections matches the pattern. Expected false, instead, it's returning true. Also account for this method.

- [ ] As highlighted in the comment below. Remove commented pieces of codes from the codebase. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))\\

Make the above-required changes and when you are ready, request another review.
Kindly drop your questions/comments :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan.

**Review by rukundoeric — APPROVED** [review]
# Status: 🍾 PROJECT APPROVED 🍾
_Excellent job on this project guys. 🎆 Keep it up._

- ### To Highlight
  - [x] The project is clean. 💯
  - [x] The project follows Git flow. 💯

- ### To Improve
  - N/A

_Congratulations on the splendid job **@team** 👍
You can now merge this PR and submit the project completion form 🚀 . I wish you all the very best in your next tasks 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/),
Slack: Rukundo Eric**

**Pull request merged**